cpp/ql/src/Likely Bugs/Underspecified Functions/TooManyArguments.ql

```diff
@@ -0,0 +1,57 @@
+/**
+ * @name Call to function with extraneous arguments
+ * @description A function call to a function passed more arguments than there are
+ *              declared parameters of the function.  This may indicate
+ *              that the code does not follow the author's intent.
+ * @kind problem
+ * @problem.severity warning
+ * @precision very-high
+ * @id cpp/futile-params
+ * @tags correctness
+ *       maintainability
+ */
+
+import cpp
+
+// True if function was ()-declared, but not (void)-declared or K&R-defined
+// or implicitly declared (i.e., lacking a prototype)
+predicate hasZeroParamDecl(Function f) {
+  exists(FunctionDeclarationEntry fde | fde = f.getADeclarationEntry() |
+    not fde.isImplicit() and
+    not fde.hasVoidParamList() and
+    fde.getNumberOfParameters() = 0 and
+    not fde.isDefinition()
+  )
+}
+
+// True if this file (or header) was compiled as a C file
+predicate isCompiledAsC(Function f) {
+  exists(File file | file.compiledAsC() |
+    file = f.getFile() or file.getAnIncludedFile+() = f.getFile()
+  )
+}
+
+predicate isWhitelisted(Function f) {
+  f instanceof BuiltInFunction
+  or
+  // The following list can be expanded as the need arises
+  exists(string name | name = f.getName() |
+    name = "static_assert" or
+    name = "_Static_assert" or
+    name = "strptime"
+  )
+}
+
+from FunctionCall fc, Function f
+where
+  f = fc.getTarget() and
+  not f.isVarargs() and
+  hasZeroParamDecl(f) and
+  isCompiledAsC(f) and
+  not isWhitelisted(f) and
```

Comment:
This manual whitelist is not a maintainable solution. Do I understand it correctly that this query will give an alert even in cases where we can't see the definition of the function? I don't think we can justify that since we have no idea what the correct argument count is. Why not require that the definition is in the snapshot? For C code, I think that's effectively `exists(f.getBlock())`.

Reply:
The purpose of the whitelist is to <i><b>not</b</i> give an alert for these functions, even though we may not see its prototype or definition.  Also, I'm not sure what to make of your <code>exists(f.getBlock())</code>.  Where would you use it and what would it accomplish?

Reply:
As far as your pointer-to-array request, I propose that we simply "unroll" the indirection one extra time, without recursion.  While it is certainly possible to write something like <code>int ***i;</code>, I don't believe we would encounter it in practice.

Reply:
Will rewrite using recursion, per 1-on-1 discussion.

Reply:
Per our (@jbj and myself) discussion, we will keep non-recursive, two-indirection solution and lower the query precision to `medium`.